crates/holochain/src/core/queue_consumer.rs

```diff
@@ -216,16 +220,17 @@ impl TriggerReceiver {
     /// Listen for one or more items to come through, draining the channel
     /// each time. Bubble up errors on empty channel.
     pub async fn listen(&mut self) -> Result<(), QueueTriggerClosedError> {
-        use tokio::sync::mpsc::error::TryRecvError;
+        use core::task::Poll;
 
         // wait for next item
-        if self.0.recv().await.is_some() {
+        if self.rx.recv().await.is_some() {
             // drain the channel
+            let mut ctx = core::task::Context::from_waker(&self.waker);
             loop {
-                match self.0.try_recv() {
-                    Err(TryRecvError::Closed) => return Err(QueueTriggerClosedError),
-                    Err(TryRecvError::Empty) => return Ok(()),
-                    Ok(()) => {}
+                match self.rx.poll_recv(&mut ctx) {
```

Comment:
I don't think these changes should have been needed - it's using a futures mpsc channel right? so the try_recv should still be there - maybe the tokio trait is shadowing the futures one?

Reply:
It seems to have been removed intentionally according to this issue: https://github.com/tokio-rs/tokio/issues/3350

After reading that thread I'm in favor of keeping the code as it is now, if we can agree that the `noop_waker` used here is appropriate for our use-case. I discussed this yesterday with @maackle and we've arrived at the conclusion that it is.

Reply:
To address the question about the source of the `mpsc` type. The code has been using the one from `tokio::sync`. I've taken a look at https://docs.rs/futures/0.3.13/futures/channel/mpsc/ and gave it a quick try to convert the code to using that. It would be a rather invasive refactor with questionable benefits.

Reply:
Yep, sorry, for some reason I thought it was already a futures channel - I'm good with this. And, yes, the noop_waker is fine, since there's no await to need rehydrating at that point in this logic.